File: Snooper/src/main/java/com/prateekj/snooper/networksnooper/helper/DataCopyHelper.java
package com.prateekj.snooper.networksnooper.helper;

import android.content.res.Resources;

import com.google.common.base.Function;
import com.prateekj.snooper.R;
import com.prateekj.snooper.formatter.ResponseFormatter;
import com.prateekj.snooper.formatter.ResponseFormatterFactory;
import com.prateekj.snooper.networksnooper.model.HttpCallRecord;
import com.prateekj.snooper.networksnooper.model.HttpHeader;
import com.prateekj.snooper.networksnooper.model.HttpHeaderValue;

import org.apache.commons.lang3.StringUtils;

import java.util.Iterator;
import java.util.List;

import static com.google.common.collect.Collections2.transform;
import static com.prateekj.snooper.networksnooper.model.HttpHeader.CONTENT_TYPE;

public class DataCopyHelper {
  private ResponseFormatterFactory responseFormatterFactory;
  private final HttpCallRecord httpCallRecord;
  private Resources resources;

  public DataCopyHelper(HttpCallRecord httpCallRecord,
                        ResponseFormatterFactory responseFormatterFactory,
                        Resources resources) {
    this.responseFormatterFactory = responseFormatterFactory;
    this.httpCallRecord = httpCallRecord;
    this.resources = resources;
  }

  public String getResponseDataForCopy() {
    return getFormattedData(httpCallRecord.getResponseHeader(CONTENT_TYPE), httpCallRecord.getResponseBody());
  }

  public String getRequestDataForCopy() {
    return getFormattedData(httpCallRecord.getRequestHeader(CONTENT_TYPE), httpCallRecord.getPayload());
  }

  public String getErrorsForCopy() {
    return httpCallRecord.getError();
  }

  public String getHeadersForCopy() {
    StringBuilder dataToCopy = new StringBuilder();
    appendHeaders(httpCallRecord.getRequestHeaders(), dataToCopy, resources.getString(R.string.request_headers));
    appendHeaders(httpCallRecord.getResponseHeaders(), dataToCopy, resources.getString(R.string.response_headers));
    return dataToCopy.toString();
  }

  public StringBuilder getHttpCallData() {
    StringBuilder dataToCopy = new StringBuilder();

    appendRequestData(dataToCopy);
    appendResponseData(dataToCopy);
    return dataToCopy;
  }

  private void appendRequestData(StringBuilder dataToCopy) {
    String formattedRequestData = getRequestDataForCopy();
    if (!StringUtils.isEmpty(formattedRequestData)) {
      dataToCopy.append(resources.getString(R.string.request_body_heading));
      dataToCopy.append("\n");
      dataToCopy.append(formattedRequestData);
    }
    appendHeaders(httpCallRecord.getRequestHeaders(), dataToCopy, resources.getString(R.string.request_headers));
  }

  private void appendHeaders(List<HttpHeader> headers, StringBuilder dataToCopy, String heading) {
    if (headers != null && !headers.isEmpty()) {
      dataToCopy.append("\n");
      dataToCopy.append(heading);
      dataToCopy.append("\n");
      for (HttpHeader header : headers) {
        dataToCopy.append(header.getName() + ": ");
        dataToCopy.append(StringUtils.join(toHeaderValues(header), ";"));
        dataToCopy.append("\n");
      }
    }
  }

  private Iterator<String> toHeaderValues(HttpHeader httpHeader) {
    return transform(httpHeader.getValues(), new Function<HttpHeaderValue, String>() {
      @Override
      public String apply(HttpHeaderValue headerValue) {
        return headerValue.getValue();
      }
    }).iterator();
  }

  private void appendResponseData(StringBuilder dataToCopy) {
    String formattedResponseData = getResponseDataForCopy();

    if (!StringUtils.isEmpty(formattedResponseData)) {
      dataToCopy.append(resources.getString(R.string.response_body_heading));
      dataToCopy.append("\n");
      dataToCopy.append(formattedResponseData);
    }
    appendHeaders(httpCallRecord.getResponseHeaders(), dataToCopy, resources.getString(R.string.response_headers));
  }


  private String getFormattedData(HttpHeader contentTypeHeader, String dataToCopy) {
    if (dataToCopy == null) {
      return "";
    }
    if (contentHeadersPresent(contentTypeHeader)) {
      ResponseFormatter formatter = this.responseFormatterFactory.getFor(contentTypeHeader.getValues().get(0).getValue());
      return formatter.format(dataToCopy);
    }
    return dataToCopy;
  }

  private boolean contentHeadersPresent(HttpHeader contentTypeHeader) {
    return contentTypeHeader != null && contentTypeHeader.getValues().size() > 0;
  }
}

File: Snooper/src/main/java/com/prateekj/snooper/networksnooper/helper/DataCopyHelper.kt
package com.prateekj.snooper.networksnooper.helper

import android.content.res.Resources
import com.prateekj.snooper.R
import com.prateekj.snooper.formatter.ResponseFormatterFactory
import com.prateekj.snooper.networksnooper.model.HttpCallRecord
import com.prateekj.snooper.networksnooper.model.HttpHeader
import com.prateekj.snooper.networksnooper.model.HttpHeader.Companion.CONTENT_TYPE

class DataCopyHelper(
  private val httpCallRecord: HttpCallRecord,
  private val responseFormatterFactory: ResponseFormatterFactory,
  private val resources: Resources
) {

  fun getResponseDataForCopy(): String {
    return getFormattedData(
      httpCallRecord.getResponseHeader(CONTENT_TYPE),
      httpCallRecord.responseBody
    )
  }

  fun getRequestDataForCopy(): String {
    return getFormattedData(
      httpCallRecord.getRequestHeader(CONTENT_TYPE),
      httpCallRecord.payload
    )

  }

  fun getErrorsForCopy(): String? = httpCallRecord.error

  fun getHeadersForCopy(): String {
    val dataToCopy = StringBuilder()
    appendHeaders(
      httpCallRecord.requestHeaders,
      dataToCopy,
      resources.getString(R.string.request_headers)
    )
    appendHeaders(
      httpCallRecord.responseHeaders,
      dataToCopy,
      resources.getString(R.string.response_headers)
    )
    return dataToCopy.toString()
  }

  fun getHttpCallData(): StringBuilder {
    val dataToCopy = StringBuilder()
    appendRequestData(dataToCopy)
    appendResponseData(dataToCopy)
    return dataToCopy
  }

  private fun appendRequestData(dataToCopy: StringBuilder) {
    val formattedRequestData = getRequestDataForCopy()
    if (formattedRequestData.isNotEmpty()) {
      val heading = resources.getString(R.string.request_body_heading)
      dataToCopy.append("$heading\n$formattedRequestData")
    }
    appendHeaders(
      httpCallRecord.requestHeaders,
      dataToCopy,
      resources.getString(R.string.request_headers)
    )
  }

  private fun appendHeaders(
    headers: List<HttpHeader>?,
    dataToCopy: StringBuilder,
    heading: String
  ) {
    if (headers != null && headers.isNotEmpty()) {
      dataToCopy.append("\n$heading\n")
      headers.forEach {
        dataToCopy.append("${it.name}: ${toHeaderValues(it)}\n")
      }
    }
  }

  private fun toHeaderValues(httpHeader: HttpHeader): String {
    return httpHeader.values.joinToString(separator = ";") { it.value }
  }

  private fun appendResponseData(dataToCopy: StringBuilder) {
    val formattedResponseData = getResponseDataForCopy()

    if (formattedResponseData.isNotEmpty()) {
      val heading = resources.getString(R.string.response_body_heading)
      dataToCopy.append("$heading\n$formattedResponseData")
    }
    appendHeaders(
      httpCallRecord.responseHeaders,
      dataToCopy,
      resources.getString(R.string.response_headers)
    )
  }


  private fun getFormattedData(contentTypeHeader: HttpHeader?, dataToCopy: String?): String {
    if (dataToCopy == null) {
      return ""
    }
    if (contentHeadersPresent(contentTypeHeader)) {
      val formatter = this.responseFormatterFactory.getFor(contentTypeHeader!!.values[0].value)
      return formatter.format(dataToCopy)
    }
    return dataToCopy
  }

  private fun contentHeadersPresent(contentTypeHeader: HttpHeader?): Boolean {
    return contentTypeHeader != null && contentTypeHeader.values.size > 0
  }
}
